src/resultsdisassemblypage.cpp

```diff
@@ -116,25 +134,73 @@ void ResultsDisassemblyPage::showDisassembly(const QString& processName, const Q
         int row = 0;
         m_model->clear();
 
-        m_model->setHorizontalHeaderLabels({tr("Assembly")});
+        QStringList headerList;
+        headerList.append({tr("Assembly")});
+        for (int i = 0; i < m_callerCalleeResults.selfCosts.numTypes(); i++) {
+            headerList.append(m_callerCalleeResults.selfCosts.typeName(i));
+        }
+        m_model->setHorizontalHeaderLabels(headerList);
 
+        int numTypes = m_callerCalleeResults.selfCosts.numTypes();
         QTextStream stream(&m_tmpFile);
         while (!stream.atEnd()) {
             QString asmLine = stream.readLine();
             if (asmLine.isEmpty() || asmLine.startsWith(QLatin1String("Disassembly"))) continue;
 
+            QStringList asmTokens = asmLine.split(QLatin1Char(':'));
+            const auto addrLine = asmTokens.value(0).trimmed();
+
+            bool ok = false;
+            QString addrLineRef = addrLine;
+            const auto addr = QStringRef(&addrLineRef).toULongLong(&ok, 16);
+            if (!ok) {
+                qWarning() << "unhandled asm line format:" << addrLine;
+                continue;
+            }
+
             QStandardItem *asmItem = new QStandardItem();
             asmItem->setText(asmLine);
             m_model->setItem(row, 0, asmItem);
+
+            // Calculate event times and add them in red to corresponding columns of the current disassembly row
+            float costLine = 0;
+            float totalCosts[numTypes];
+            auto &entry = m_callerCalleeResults.entry(m_curSymbol);
+
+            QHash<quint64, Data::LocationCost>::iterator i = entry.offsetMap.begin();
+            while (i != entry.offsetMap.end()) {
+                quint64 relAddr = i.key();
+                Data::LocationCost locationCost = i.value();
+                for (int event = 0; event < numTypes; event++) {
+                    float cost = locationCost.selfCost[event];
+                    totalCosts[event] += cost;
+                }
+                i++;
+            }
+
+            auto it = entry.offsetMap.find(addr);
+            if (it != entry.offsetMap.end()) {
+                Data::LocationCost locationCost = it.value();
+                for (int event = 0; event < numTypes; event++) {
+                    costLine = locationCost.selfCost[event];
+                    QString costInstruction = Util::formatCostRelative(costLine, totalCosts[event], true);
+
+                    //FIXME QStandardItem stuff should be reimplemented properly
+                    QStandardItem *costItem = new QStandardItem(costInstruction);
+                    costItem->setData(costLine, CostRole);
+                    costItem->setData(totalCosts[event], TotalCostRole);
```

Comment:
this would now actually compute the total cost relative to the function, which is imo pretty confusing

it would be much better to use the overall total cost instead:

costItem->setData(m_callerCalleeResults->totalCost(event), TotalCostRole);

Reply:
Why do you want relative-to-total instead of relative-to-functions numbers? At least 'perf annotate' computes them relative-to-function. I think we should follow this approach and it should be relative-to-function. If you want relative-to-total values,  probably they can be computed separately and shown nearby column. I can write TODO about them.

Reply:
Because the rest of hotspot also presents the data in this way. It would be very confusing if we suddenly show the data differently in this one new view.

and yes, I insist on this change. unification is important